File: connection/src/main/java/otea/connection/controller/SubAmbitsController.java
package otea.connection.controller;

import java.io.IOException;
import java.util.List;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

import cli.indicators.SubAmbit;
import otea.connection.ConnectionClient;
import otea.connection.api.SubAmbitsApi;
import retrofit2.Call;
import retrofit2.Response;

/**
 * Controller class for subAmbit operations
 *
 * @author Pablo Ahíta del Barrio
 * @version 1
 * */
public class SubAmbitsController {

    /**SubAmbits api to connect to the server*/
    private static SubAmbitsApi api;

    /**SubAmbitsController instance for singleton*/
    private static SubAmbitsController instance;

    /**Class constructor*/
    private SubAmbitsController(){
        api= ConnectionClient.getInstance().getRetrofit().create(SubAmbitsApi.class);
    }

    /**
     * Synchronized method that obtains the instance
     *
     * @return Instance of subAmbit
     * */
    public static SubAmbitsController getInstance(){
        if(instance==null){
            synchronized (AmbitsController.class){
                if(instance==null){
                    instance=new SubAmbitsController();
                }
            }
        }
        return instance;
    }

    /**
     * Method that obtains from the database all the subAmbits
     *
     * @return All subAmbits
     * */
    public static List<SubAmbit> GetAll(){
        ExecutorService executor = Executors.newSingleThreadExecutor();
        Callable<List<SubAmbit>> callable = new Callable<List<SubAmbit>>() {
            @Override
            public List<SubAmbit> call() throws Exception {
                Call<List<SubAmbit>> call = api.GetAll();
                Response<List<SubAmbit>> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };
        try {
            Future<List<SubAmbit>> future = executor.submit(callable);
            List<SubAmbit> list = future.get();
            executor.shutdown();
            return list;
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Method that obtains from the database all the subAmbits of an ambit
     *
     * @param idAmbit - Ambit identifier
     * @return All subAmbits
     * */
    public static List<SubAmbit> GetAllByAmbit(int idAmbit){
        ExecutorService executor = Executors.newSingleThreadExecutor();
        Callable<List<SubAmbit>> callable = new Callable<List<SubAmbit>>() {
            @Override
            public List<SubAmbit> call() throws Exception {
                Call<List<SubAmbit>> call = api.GetAllByAmbit(idAmbit);
                Response<List<SubAmbit>> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };
        try {
            Future<List<SubAmbit>> future = executor.submit(callable);
            List<SubAmbit> list = future.get();
            executor.shutdown();
            return list;
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Method that obtains from the database a subAmbit using its identifier
     *
     * @param id - subAmbit identifier
     * @return The subAmbit if exists, null if not.
     * */

    public static SubAmbit Get(int id){
        ExecutorService executor = Executors.newSingleThreadExecutor();
        Callable<SubAmbit> callable = new Callable<SubAmbit>() {
            @Override
            public SubAmbit call() throws Exception {
                Call<SubAmbit> call = api.Get(id);
                Response<SubAmbit> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };
        try {
            Future<SubAmbit> future = executor.submit(callable);
            SubAmbit result = future.get();
            executor.shutdown();
            return result;
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Method that appends a new subAmbit to the database
     *
     * @param subAmbit - SubAmbit to append
     * @return SubAmbit append, null if not
     * */
    public static SubAmbit Create(SubAmbit subAmbit){
        ExecutorService executor = Executors.newSingleThreadExecutor();
        Callable<SubAmbit> callable = new Callable<SubAmbit>() {
            @Override
            public SubAmbit call() throws Exception {
                Call<SubAmbit> call = api.Create(subAmbit);
                Response<SubAmbit> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };
        try {
            Future<SubAmbit> future = executor.submit(callable);
            SubAmbit result = future.get();
            executor.shutdown();
            return result;
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Method that updates the subAmbit
     *
     * @param idSubAmbit - SubAmbit identifier
     * @param subAmbit - new subAmbit
     * @return subAmbit if was updated, null if not
     *
     * */
    public static SubAmbit Update(int idSubAmbit, SubAmbit subAmbit){
        ExecutorService executor = Executors.newSingleThreadExecutor();
        Callable<SubAmbit> callable = new Callable<SubAmbit>() {
            @Override
            public SubAmbit call() throws Exception {
                Call<SubAmbit> call = api.Update(idSubAmbit,subAmbit);
                Response<SubAmbit> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };
        try {
            Future<SubAmbit> future = executor.submit(callable);
            SubAmbit result = future.get();
            executor.shutdown();
            return result;
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Method that deletes the subAmbit using its identifier
     *
     * @param id - SubAmbit identifier
     * @return Deleted subAmbit if deletion was successful, null if not
     * */
    public static SubAmbit Delete(int id){
        ExecutorService executor = Executors.newSingleThreadExecutor();
        Callable<SubAmbit> callable = new Callable<SubAmbit>() {
            @Override
            public SubAmbit call() throws Exception {
                Call<SubAmbit> call = api.Delete(id);
                Response<SubAmbit> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };
        try {
            Future<SubAmbit> future = executor.submit(callable);
            SubAmbit result = future.get();
            executor.shutdown();
            return result;
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
    }
}

File: OTEAServer/Controllers/SubAmbitsController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using OTEAServer.Misc;
using OTEAServer.Models;

namespace OTEAServer.Controllers
{

    /// <summary>
    /// Controller class for subambits operations
    /// Author: Pablo Ahíta del Barrio
    /// Version: 1
    /// </summary>
    [ApiController]
    [Route("SubAmbits")]
    public class SubAmbitsController : ControllerBase
    {
        /// <summary>
        /// Database context
        /// </summary>
        private readonly DatabaseContext _context;

        /// <summary>
        /// Class constructor
        /// </summary>
        /// <param name="context">Database context</param>
        public SubAmbitsController(DatabaseContext context)
        {
            _context = context;
        }

        /// <summary>
        /// Method that obtains from the database all the subAmbits
        /// </summary>
        /// <returns>All subAmbits</returns>
        [HttpGet("all")]
        [Authorize]
        public IActionResult GetAll([FromHeader] string Authorization)
        {
            try
            {
                var subAmbits = _context.SubAmbits.ToList();
                return Ok(subAmbits);
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }

        /// <summary>
        /// Method that obtains from the database all the subAmbits
        /// </summary>
        /// <param name="idAmbit">Ambit identifier</param>
        /// <returns>All subAmbits</returns>
        [HttpGet("allByAmbit")]
        [Authorize]
        public IActionResult GetAllByAmbit(int idAmbit, [FromHeader] string Authorization)
        {
            try
            {
                var subAmbits = _context.SubAmbits.Where(s => s.idAmbit == idAmbit).ToList();
                return Ok(subAmbits);
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }

        /// <summary>
        /// Method that obtains from the database a subAmbit using its identifier
        /// </summary>
        /// <param name="id">subAmbit identifier</param>
        /// <returns>The subAmbit if exists, null if not.</returns>
        [HttpGet("get")]
        [Authorize]
        public ActionResult<SubAmbit> Get([FromQuery] int id, [FromHeader] string Authorization)
        {
            try
            {
                var subAmbit = _context.SubAmbits.FirstOrDefault(a => a.idSubAmbit == id);

                if (subAmbit == null)
                    return NotFound();

                return subAmbit;
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }

        /// <summary>
        /// Method that appends a new subAmbit to the database
        /// </summary>
        /// <param name="subAmbit">SubAmbit to append</param>
        /// <returns>SubAmbit append, null if not</returns>
        [HttpPost]
        [Authorize(Policy = "Administrator")]
        public IActionResult Create([FromBody] SubAmbit subAmbit, [FromHeader] string Authorization)
        {
            try
            {
                _context.SubAmbits.Add(subAmbit);
                _context.SaveChanges();
                return CreatedAtAction(nameof(Get), new { idSubAmbit = subAmbit.idSubAmbit }, subAmbit);
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }


        /// <summary>
        /// Method that updates the subAmbit
        /// </summary>
        /// <param name="idSubAmbit">SubAmbit identifier</param>
        /// <param name="subAmbit">new subAmbit</param>
        /// <returns>subAmbit if was updated, null if not</returns>
        [HttpPut]
        [Authorize(Policy = "Administrator")]
        public IActionResult Update([FromQuery] int idSubAmbit, [FromBody] SubAmbit subAmbit, [FromHeader] string Authorization)
        {
            try
            {
                if (idSubAmbit != subAmbit.idSubAmbit)
                    return BadRequest();

                var existingSubAmbit = _context.SubAmbits.FirstOrDefault(a => a.idSubAmbit == idSubAmbit);
                if (existingSubAmbit is null)
                    return NotFound();


                existingSubAmbit.idSubAmbit = subAmbit.idSubAmbit;
                existingSubAmbit.idAmbit = subAmbit.idAmbit;
                existingSubAmbit.descriptionEnglish = subAmbit.descriptionEnglish;
                existingSubAmbit.descriptionSpanish = subAmbit.descriptionSpanish;
                existingSubAmbit.descriptionFrench = subAmbit.descriptionFrench;
                existingSubAmbit.descriptionBasque = subAmbit.descriptionBasque;
                existingSubAmbit.descriptionCatalan = subAmbit.descriptionCatalan;
                existingSubAmbit.descriptionDutch = subAmbit.descriptionDutch;
                existingSubAmbit.descriptionGalician = subAmbit.descriptionGalician;
                existingSubAmbit.descriptionGerman = subAmbit.descriptionGerman;
                existingSubAmbit.descriptionItalian = subAmbit.descriptionItalian;
                existingSubAmbit.descriptionPortuguese = subAmbit.descriptionPortuguese;
                _context.SaveChanges();

                return Ok(existingSubAmbit);
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }

        /// <summary>
        /// Method that deletes the subAmbit using its identifier
        /// </summary>
        /// <param name="id">SubAmbit identifier</param>
        /// <returns>Deleted subAmbit if deletion was successful, null if not</returns>
        [HttpDelete]
        [Authorize(Policy = "Administrator")]
        public IActionResult Delete([FromQuery] int id, [FromHeader] string Authorization)
        {
            try
            {
                var subAmbit = _context.SubAmbits.FirstOrDefault(a => a.idSubAmbit == id);

                if (subAmbit is null)
                    return NotFound();

                _context.SubAmbits.Remove(subAmbit);
                _context.SaveChanges();

                return NoContent();
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }
    }
}
